pkg/scheduler/api/job_info.go

```diff
@@ -723,6 +723,27 @@ func (ji *JobInfo) PendingBestEffortTaskNum() int32 {
 	return int32(count)
 }
 
+// FitFailedRoles returns the job roles' failed fit records
+func (ji *JobInfo) FitFailedRoles() map[string]struct{} {
+	failedRoles := map[string]struct{}{}
+	for tid := range ji.NodesFitErrors {
+		task := ji.Tasks[tid]
+		failedRoles[task.TaskSpec] = struct{}{}
+	}
+	return failedRoles
+}
+
+// TaskHasFitErrors checks if the task has fit errors and can continue try predicating
+func (ji *JobInfo) TaskHasFitErrors(task *TaskInfo) bool {
+	// if the task didn't set the spec key, should not use the cache
+	if len(task.TaskSpec) == 0 {
+		return false
+	}
+
+	_, exist := ji.FitFailedRoles()[task.TaskSpec]
```

Comment:
`task.TaskSpec` is just set in vcjob, and other case will be "", can this take efffect?

Reply:
Will retun false in the if branch.

Reply:
Actually, the empty spec and default to enable cache will cause job can not be scheduled in some corner case. See issue https://github.com/volcano-sh/volcano/issues/3527. We'd better don't enable cache when spec is empty in predicate helper.